xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+  private boolean initialStarted;
+  private ScheduledHandle restartTimer;
+  private ObjectPool<XdsClient> xdsClientPool;
+  private XdsClient xdsClient;
+  private DiscoveryState discoveryState;
+  private volatile Server delegate;
+
+  XdsServerWrapper(
+      String listenerAddress,
+      ServerBuilder<?> delegateBuilder,
+      long retryDelayNano,
+      XdsServingStatusListener listener,
+      AtomicReference<FilterChainSelector> filterChainSelectorRef,
+      XdsClientPoolFactory xdsClientPoolFactory) {
+    this.listenerAddress = checkNotNull(listenerAddress, "listenerAddress");
+    this.delegateBuilder = checkNotNull(delegateBuilder, "delegateBuilder");
+    this.retryDelayNano = retryDelayNano;
+    this.listener = checkNotNull(listener, "listener");
+    this.filterChainSelectorRef = checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+    this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.delegate = delegateBuilder.build();
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    syncContext.drain();
+    IOException exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw exception;
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException());
+      initialStartFuture.set(null);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
+    //  from bootstrap.
+    boolean useProtocolV3 = xdsClient.getBootstrapInfo().getServers().get(0).isUseProtocolV3();
+    String listenerTemplate = xdsClient.getBootstrapInfo().getServerListenerResourceNameTemplate();
+    if (!useProtocolV3 || listenerTemplate == null) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException());
+      initialStartFuture.set(null);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        delegate.shutdown();
```

Comment:
Is there a reason for this order? Why not shut down the delegate first and then shut down the xdsClient (or stop the listener subscription)? If the caller has requested this XdsServer's shutdown they probably want the incoming calls to stop to this server as soon as possible even if the listener unsubscribe is delayed a little bit

Reply:
> Shutdown should immediately stop accepting new RPCs but allow the old ones to finish. Current order is for this consideration.

Doesn't make sense. `delegate.shutdown()` itself achieves that (allow old RPCs to finish) and not the order here.   I didn't understand the relevance of the rest of your comments

Reply:
> delegate.shutdown() itself achieves that (allow old RPCs to finish) 

It does allow old RPCs to finish, but it does not reject new RPCs immediately, because `delegate` `shutdown()` , or eventually [netty server shutdown](https://github.com/grpc/grpc-java/blob/d4fa0ecc07495097453b0a2848765f076b9e714c/netty/src/main/java/io/grpc/netty/NettyServer.java#L361), won't reject coming connections, it processes new RPCs until the last one.  So it may never shutdown.
Instead we should close the new connection immediately by clearing filterChainSelector first.

Reply:
> The doc for Server.shutdown() says "Initiates an orderly shutdown in which preexisting calls continue but new calls are rejected."

Good find. I was mistaken, netty does close the server channel immediately.
Regardless could you remind me what was not appropriate with the current implementation?

Reply:
My original suggestion was to reverse the order:

```
delegate.shutdown();
internalShutdown();
```
because I believe this is strictly better than the current one.